### PR TITLE
fix: `to_gbq` allows strings for DATE and floats for NUMERIC, require pandas 0.24+ and db-dtypes

### LINKED ISSUE #421: ConversionError: Could not convert DataFrame to Parquet. | After upgrate to 0.16.0
#### Environment details

  - OS type and version: Windows 10 x64
  - Python version: 3.8.5
  - pip version: 20.2.4
  - `pandas-gbq` version: 0.16.0

#### Steps to reproduce

This code was executed in previous version of pandas-gbq (0.15.0) and was successfully executed.

#### Code example

```python

import os
import pandas as pd
import pandas_gbq as gbq

table_schema = [{
    "name": "id", 
    "type": "INTEGER"
},{
    "name": "nombre", 
    "type": "STRING"
},{
    "name": "precio", 
    "type": "NUMERIC"
},{
    "name": "fecha", 
    "type": "DATE"
}]

data = pd.DataFrame({'id': [123],'nombre': ['Anderson'],'precio': [1.25],'fecha': ['2021-12-12']})

project_id = 'proyecto111'
table_name = 'prueba.clientes'

gbq.to_gbq(data, table_name, project_id, if_exists='append', table_schema = table_schema)

```

#### Stack trace
```
---------------------------------------------------------------------------
ArrowInvalid                              Traceback (most recent call last)
~\anaconda3\lib\site-packages\pandas_gbq\load.py in load_parquet(client, dataframe, destination_table_ref, location, schema)
     74     try:
---> 75         client.load_table_from_dataframe(
     76             dataframe, destination_table_ref, job_config=job_config, location=location,

~\anaconda3\lib\site-packages\google\cloud\bigquery\client.py in load_table_from_dataframe(self, dataframe, destination, num_retries, job_id, job_id_prefix, location, project, job_config, parquet_compression, timeout)
   2650 
-> 2651                     _pandas_helpers.dataframe_to_parquet(
   2652                         dataframe,

~\anaconda3\lib\site-packages\google\cloud\bigquery\_pandas_helpers.py in dataframe_to_parquet(dataframe, bq_schema, filepath, parquet_compression, parquet_use_compliant_nested_type)
    585     bq_schema = schema._to_schema_fields(bq_schema)
--> 586     arrow_table = dataframe_to_arrow(dataframe, bq_schema)
    587     pyarrow.parquet.write_table(

~\anaconda3\lib\site-packages\google\cloud\bigquery\_pandas_helpers.py in dataframe_to_arrow(dataframe, bq_schema)
    528         arrow_arrays.append(
--> 529             bq_to_arrow_array(get_column_or_index(dataframe, bq_field.name), bq_field)
    530         )

~\anaconda3\lib\site-packages\google\cloud\bigquery\_pandas_helpers.py in bq_to_arrow_array(series, bq_field)
    289         return pyarrow.StructArray.from_pandas(series, type=arrow_type)
--> 290     return pyarrow.Array.from_pandas(series, type=arrow_type)
    291 

~\anaconda3\lib\site-packages\pyarrow\array.pxi in pyarrow.lib.Array.from_pandas()

~\anaconda3\lib\site-packages\pyarrow\array.pxi in pyarrow.lib.array()

~\anaconda3\lib\site-packages\pyarrow\array.pxi in pyarrow.lib._ndarray_to_array()

~\anaconda3\lib\site-packages\pyarrow\error.pxi in pyarrow.lib.check_status()

ArrowInvalid: Got bytestring of length 8 (expected 16)

The above exception was the direct cause of the following exception:

ConversionError                           Traceback (most recent call last)
<ipython-input-5-49cafceeee1e> in <module>
     24 table_name = 'prueba.clientes'
     25 
---> 26 gbq.to_gbq(data, table_name, project_id, if_exists='append', table_schema = table_schema)

~\anaconda3\lib\site-packages\pandas_gbq\gbq.py in to_gbq(dataframe, destination_table, project_id, chunksize, reauth, if_exists, auth_local_webserver, table_schema, location, progress_bar, credentials, api_method, verbose, private_key)
   1093         return
   1094 
-> 1095     connector.load_data(
   1096         dataframe,
   1097         destination_table_ref,

~\anaconda3\lib\site-packages\pandas_gbq\gbq.py in load_data(self, dataframe, destination_table_ref, chunksize, schema, progress_bar, api_method)
    544 
    545         try:
--> 546             chunks = load.load_chunks(
    547                 self.client,
    548                 dataframe,

~\anaconda3\lib\site-packages\pandas_gbq\load.py in load_chunks(client, dataframe, destination_table_ref, chunksize, schema, location, api_method)
    164 ):
    165     if api_method == "load_parquet":
--> 166         load_parquet(client, dataframe, destination_table_ref, location, schema)
    167         # TODO: yield progress depending on result() with timeout
    168         return [0]

~\anaconda3\lib\site-packages\pandas_gbq\load.py in load_parquet(client, dataframe, destination_table_ref, location, schema)
     77         ).result()
     78     except pyarrow.lib.ArrowInvalid as exc:
---> 79         raise exceptions.ConversionError(
     80             "Could not convert DataFrame to Parquet."
     81         ) from exc

ConversionError: Could not convert DataFrame to Parquet.
```

### PR DESCRIPTION
**Pull request by tswast**
deps: require pandas 0.24+ and db-dtypes for TIME/DATE extension dtypes

~Review https://github.com/googleapis/python-bigquery-pandas/pull/420 first! This PR is based on changes to the system tests introduced there.~

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-pandas/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes googleapis/python-bigquery-pandas#421 🦕

### CODE REVIEW

**Comment by tswast** [comment]
Need to package db-dtypes package in order to pass conda session.

File: tests/system/test_to_gbq.py [hunk 1]
```diff
@@ -118,53 +120,89 @@ def test_series_round_trip(
...
         (
             pandas.DataFrame(
                 {
+                    "row_num": [0, 1, 2],
                     "date_col": pandas.Series(
                         ["2021-04-17", "1999-12-31", "2038-01-19"], dtype="dbdate",
                     ),
                 }
             ),
+            None,
```
Comment (plamut) [inline tests/system/test_to_gbq.py]: Why is `expected_df` None here? I'm probably missing something.

**Edit:** I see below, it's implicitly saying that it should bee the same as `input_df`.

We might want to consider making these parametrized cases more understandable on the first pass already. Two ideas:
 - Using namedtuples, giving the tuple elements names through the kwarg on nametuple creation.
 - Adding an explanatory comment next to the constants (`None`, `False`...) used in the branching inside the test. Right now one needs to scroll down into the test body to make sense of them.

File: tests/unit/test_load.py [hunk 2]
```diff
@@ -137,3 +141,109 @@ def test_load_chunks_omits_policy_tags(
...
+                ],
+                dtype="object",
+            ),
+            "row_num_2": [0, 1, 2],
+        },
+        # Use multiple columns to ensure column order is maintained.
+        columns=["row_num", "repeated_col", "row_num_2"],
+    )
+    expected = dataframe.copy()
+    schema = {"fields": [{"name": "date_col", "type": "DATE", "mode": "REPEATED"}]}
```
Comment (plamut) [inline tests/unit/test_load.py:247]: Copy/paste error? The name of the field should be `"repeated_col"` it seems, the name mismatch is not intentional?

File: owlbot.py [hunk 3]
```diff
@@ -32,7 +32,7 @@
 templated_files = common.py_library(
     unit_test_python_versions=["3.7", "3.8", "3.9", "3.10"],
     system_test_python_versions=["3.7", "3.8", "3.9", "3.10"],
-    cov_level=86,
+    cov_level=88,
```
Comment (plamut) [inline owlbot.py:35]: Any way to get rid of the magic threshold quickly? Probably deserves a separate PR?

**Review by plamut — COMMENTED** [review]
Found a few things worth double-checking, but the general picture looks good.

File: owlbot.py (hunk 3 above)
Comment (tswast) [reply owlbot.py:35]: I've got https://github.com/googleapis/python-bigquery-pandas/issues/392 open to bump us up to 100%. 88% is a lot better than what it used to be...

File: pandas_gbq/load.py [hunk 4]
```diff
@@ -56,6 +58,47 @@ def split_dataframe(dataframe, chunksize=None):
...
 
+def cast_dataframe_for_parquet(
+    dataframe: pandas.DataFrame, schema: Optional[Dict[str, Any]],
+) -> pandas.DataFrame:
+    """Cast columns to needed dtype when writing parquet files.
+
+    See: https://github.com/googleapis/python-bigquery-pandas/issues/421
+    """
+    columns = schema.get("fields", [])
+    for column in columns:
```
[the remark this answers is not in the document]
Comment (tswast) [reply pandas_gbq/load.py:69]: Yeah, probably a case we should cover.

File: tests/system/test_to_gbq.py (hunk 1 above)
Comment (tswast) [reply tests/system/test_to_gbq.py]: I like the namedtuple idea

**Comment by tswast** [comment]
```
nox > Running session system-3.7
nox > Creating virtual environment (virtualenv) using python3.7 in .nox/system-3-7
nox > python -m pip install --pre grpcio
nox > python -m pip install mock pytest google-cloud-testutils -c /tmpfs/src/github/python-bigquery-pandas/testing/constraints-3.7.txt
nox > python -m pip install -e .[tqdm] -c /tmpfs/src/github/python-bigquery-pandas/testing/constraints-3.7.txt
nox > py.test --quiet --junitxml=system_3.7_sponge_log.xml tests/system
.sss......F.FF.........................................................s [ 73%]
..........................                                               [100%]
=================================== FAILURES ===================================
_____ TestReadGBQIntegration.test_should_properly_handle_nullable_integers _____

self = <system.test_gbq.TestReadGBQIntegration object at 0x7f550a5fe910>
project_id = 'precise-truck-742'

    def test_should_properly_handle_nullable_integers(self, project_id):
        if PANDAS_VERSION < NULLABLE_INT_PANDAS_VERSION:
            pytest.skip(msg=NULLABLE_INT_MESSAGE)
    
        query = """SELECT * FROM
                    UNNEST([1, NULL]) AS nullable_integer
                """
        df = gbq.read_gbq(
            query,
            project_id=project_id,
            credentials=self.credentials,
            dialect="standard",
            dtypes={"nullable_integer": "Int64"},
        )
        tm.assert_frame_equal(
            df,
            DataFrame(
>               {"nullable_integer": pandas.Series([1, pandas.NA], dtype="Int64")}
            ),
        )
E       AttributeError: module 'pandas' has no attribute 'NA'

tests/system/test_gbq.py:192: AttributeError
----------------------------- Captured stderr call -----------------------------

Downloading: 0rows [00:00, ?rows/s]
Downloading: 100%|██████████| 2/2 [00:00<00:00,  5.88rows/s]
______ TestReadGBQIntegration.test_should_properly_handle_nullable_longs _______

self = <system.test_gbq.TestReadGBQIntegration object at 0x7f550a580bd0>
project_id = 'precise-truck-742'

    def test_should_properly_handle_nullable_longs(self, project_id):
        if PANDAS_VERSION < NULLABLE_INT_PANDAS_VERSION:
            pytest.skip(msg=NULLABLE_INT_MESSAGE)
    
        query = """SELECT * FROM
                    UNNEST([1 << 62, NULL]) AS nullable_long
                """
        df = gbq.read_gbq(
            query,
            project_id=project_id,
            credentials=self.credentials,
            dialect="standard",
            dtypes={"nullable_long": "Int64"},
        )
        tm.assert_frame_equal(
            df,
            DataFrame(
>               {"nullable_long": pandas.Series([1 << 62, pandas.NA], dtype="Int64")}
            ),
        )
E       AttributeError: module 'pandas' has no attribute 'NA'

tests/system/test_gbq.py:223: AttributeError
----------------------------- Captured stderr call -----------------------------

Downloading: 0rows [00:00, ?rows/s]
Downloading: 100%|██████████| 2/2 [00:00<00:00, 10.43rows/s]
_______ TestReadGBQIntegration.test_should_properly_handle_null_integers _______

self = <system.test_gbq.TestReadGBQIntegration object at 0x7f550a5c5850>
project_id = 'precise-truck-742'

    def test_should_properly_handle_null_integers(self, project_id):
        if PANDAS_VERSION < NULLABLE_INT_PANDAS_VERSION:
            pytest.skip(msg=NULLABLE_INT_MESSAGE)
    
        query = "SELECT CAST(NULL AS INT64) AS null_integer"
        df = gbq.read_gbq(
            query,
            project_id=project_id,
            credentials=self.credentials,
            dialect="standard",
            dtypes={"null_integer": "Int64"},
        )
        tm.assert_frame_equal(
>           df, DataFrame({"null_integer": pandas.Series([pandas.NA], dtype="Int64")}),
        )
E       AttributeError: module 'pandas' has no attribute 'NA'

tests/system/test_gbq.py:240: AttributeError
----------------------------- Captured stderr call -----------------------------
```

Possible we need to adjust NULLABLE_INT_PANDAS_VERSION

**Comment by tswast** [comment]
Marking as DO NOT MERGE as a reminder to make sure `deps: require pandas 0.24+ and db-dtypes for TIME/DATE extension dtypes` is also included as a footer to the PR, as I'd like to see if we can use this feature: https://github.com/googleapis/release-please/pull/686

**Comment by tswast** [comment]
Per https://github.com/googleapis/release-please/issues/821, the extra metadata isn't important. It's that the extra commits are listed as a footer to the commit message without anything in-between.

**Comment by tswast** [comment]
Ready for (re)review.

**Review by plamut — APPROVED** [review]
LGTM now, thanks for the cleanup!

**Comment by plamut** [comment]
The nightly test failure is probably flakiness? The logs say "killed"?

**Comment by tswast** [comment]
Nightly failure isn't even flakey. It always fails because the conda package installation times out. https://github.com/googleapis/python-bigquery-pandas/issues/424

Switching to `mamba` for installs might help.
